gamemodes/terrortown/entities/entities/ttt_basegrenade_proj.lua

```diff
@@ -0,0 +1,170 @@
+---
+-- common grenade projectile code
+-- @class ENT
+-- @section ttt_basegrenade_proj
+
+AddCSLuaFile()
+
+ENT.Type = "anim"
+ENT.Model = Model("models/weapons/w_eq_flashbang_thrown.mdl")
+ENT.playerDropWeaponOnContact = false
+ENT.sphericalPhysicsRadius = 0
+ENT.damageTriggerProjectileExplosion = true
+ENT.preventDamageChainReaction = true
+ENT.bounceFactor = 0
+ENT.bounceMomentumLoss = 0
+ENT.impactSoundSpeed = 50
+ENT.throwSound = Sound("entities/entities/ttt_basegrenade_proj/toss.wav")
+ENT.impactSound = Sound("entities/entities/ttt_basegrenade_proj/impact.wav")
+
+---
+-- @realm shared
+function ENT:SetupDataTables()
+	self:NetworkVar("Float", 0, "ExplodeTime")
+	self:NetworkVar("Bool", 0, "ShotToActivate")
+	self:NetworkVar("Bool", 1, "FailedToActivate")
+	self:NetworkVar("Bool", 2, "Exploded")
+	self:NetworkVar("String", 0, "LaunchWeaponClass")
+	self:NetworkVar("Entity", 0, "Thrower")
+
+	if SERVER then
+		self:SetExplodeTime(0)
+		self:SetFailedToActivate(false)
+		self:SetExploded(false)
+		self:SetShotToActivate(false)
+	end
+end
+
+---
+-- @realm shared
+function ENT:Initialize()
+	self:AddEFlags( EFL_FORCE_CHECK_TRANSMIT )
+	self:SetModel(self.Model)
+
+	if self.sphericalPhysicsRadius > 0 then
+		self:PhysicsInitSphere(self.sphericalPhysicsRadius, "metal")
+		self:SetMoveType( MOVETYPE_VPHYSICS )
+		self:SetSolid(SOLID_VPHYSICS)
+	else
+		self:PhysicsInit(SOLID_VPHYSICS)
+		self:SetMoveType(MOVETYPE_VPHYSICS)
+		self:SetSolid(SOLID_BBOX)
+	end
+	self:SetCollisionGroup(COLLISION_GROUP_NONE)
+end
+
+---
+-- Everything that needs cleaning up prior to removal, but OnRemove is too late for.
+-- @realm shared
+function ENT:Destroy()
+	if CLIENT then return end
+	self:SetExploded(true)
+	self:SetDetonateExact(0)
+	self:AddEFlags(EFL_FORCE_CHECK_TRANSMIT)
+	self:AddEFlags(EFL_DORMANT)
+	if SERVER then self:Remove() end
+end
+
+---
+-- @param number length How far away _from now_ to detonate.
+-- @realm shared
+function ENT:SetDetonateTimer(length)
+	self:SetDetonateExact( CurTime() + length )
+end
+
+---
+-- @param nil|number t The time we should explode, or <code>CurTime()</code> if falsy.
+-- @realm shared
+function ENT:SetDetonateExact(t)
+	self:SetExplodeTime(t or CurTime())
```

Comment:
Does this need to run on the client? Setting NetworkVars on the client is discouraged so i'm curious if we could simply guard against setting this on the client and be fine.
```suggestion
	if CLIENT then return end
	self:SetExplodeTime(t or CurTime())
```